finality-aleph/src/validator_network/protocols/v1/mod.rs

```diff
@@ -0,0 +1,446 @@
+            },
+            _ => Heartbeat,
+        };
+        sender = send_data(sender, to_send).await?;
+    }
+}
+
+async fn receiving<D: Data, S: AsyncRead + Unpin + Send>(
+    mut stream: S,
+    data_for_user: mpsc::UnboundedSender<D>,
+) -> Result<(), ProtocolError> {
+    use Message::*;
+    loop {
+        let (old_stream, message) = timeout(
+            MAX_MISSED_HEARTBEATS * HEARTBEAT_TIMEOUT,
+            receive_data(stream),
+        )
+        .await
+        .map_err(|_| ProtocolError::CardiacArrest)??;
+        stream = old_stream;
+        match message {
+            Data(data) => data_for_user
+                .unbounded_send(data)
+                .map_err(|_| ProtocolError::NoUserConnection)?,
+            Heartbeat => (),
+        }
+    }
+}
+
+async fn manage_connection<D: Data, S: AsyncWrite + Unpin + Send, R: AsyncRead + Unpin + Send>(
+    sender: S,
+    receiver: R,
+    data_from_user: mpsc::UnboundedReceiver<D>,
+    data_for_user: mpsc::UnboundedSender<D>,
+) -> Result<(), ProtocolError> {
+    let sending = sending(sender, data_from_user);
+    let receiving = receiving(receiver, data_for_user);
+    tokio::select! {
+        result = receiving => result,
+        result = sending => result,
+    }
+}
+
+/// Performs the outgoing handshake, and then manages a connection sending and receiving data.
+/// Exits on parent request, or in case of broken or dead network connection.
+pub async fn outgoing<D: Data, S: Splittable>(
+    stream: S,
+    authority_pen: AuthorityPen,
+    peer_id: AuthorityId,
+    result_for_parent: mpsc::UnboundedSender<ResultForService<D>>,
+    data_for_user: mpsc::UnboundedSender<D>,
+) -> Result<(), ProtocolError> {
+    trace!(target: "validator-network", "Extending hand to {}.", peer_id);
+    let (sender, receiver) = v0_handshake_outgoing(stream, authority_pen, peer_id.clone()).await?;
+    info!(target: "validator-network", "Outgoing handshake with {} finished successfully.", peer_id);
+    let (data_for_network, data_from_user) = mpsc::unbounded();
+    result_for_parent
+        .unbounded_send((peer_id.clone(), Some(data_for_network), ConnectionType::New))
+        .map_err(|_| ProtocolError::NoParentConnection)?;
+    debug!(target: "validator-network", "Starting worker for communicating with {}.", peer_id);
+    manage_connection(sender, receiver, data_from_user, data_for_user).await
+}
+
+/// Performs the incoming handshake, and then manages a connection sending and receiving data.
+/// Exits on parent request, or in case of broken or dead network connection.
+pub async fn incoming<D: Data, S: Splittable>(
+    stream: S,
+    authority_pen: AuthorityPen,
+    result_for_parent: mpsc::UnboundedSender<ResultForService<D>>,
+    data_for_user: mpsc::UnboundedSender<D>,
+) -> Result<(), ProtocolError> {
+    trace!(target: "validator-network", "Waiting for extended hand...");
+    let (sender, receiver, peer_id) = v0_handshake_incoming(stream, authority_pen).await?;
```

Comment:
I suppose it's a temp mock, but I would suggest adding a few lines of real mock, because it's really confusing xD

Reply:
Huh? Why do you think this is a temp mock, or any mock at all?

Reply:
oh, I assumed that vx uses things from vx and `v1_handshake` will be added soon

Reply:
Handshake versioning is separate from protocol versioning, and both v1 and v0 use handshake v0 – there was no reason to write a new handshake. That's also why `handshake.rs` is in `protocol/` directly.

Reply:
uhm, then the name `v0_handshake` is a bit confusing :P

Reply:
Kinda, I expect it will become less confusing with time, as it becomes clearly unrelated to the protocol versioning, as it will never match the protocol version again. It might also become clearer when there are multiple handshakes – they would likely end up in places like `handshake::vX`.